api/views/share.py

```diff
@@ -59,102 +61,34 @@ def rawdata(self, request, pk=None):
     # required_scopes = ['', ]
     filter_class = ShareObjectFilterSet
 
+    # override to convert encoded pk to an actual pk
+    def get_object(self):
+        queryset = self.filter_queryset(self.get_queryset())
 
-def make_creative_work_view_set_class(model):
-    class CreativeWorkViewSet(ShareObjectViewSet):
-        serializer_class = serializers.make_creative_work_serializer_class(model)
-        queryset = serializer_class.Meta.model.objects.all().select_related('extra')
-    return CreativeWorkViewSet
-
-
-# Creative work attributes
-
-class TagViewSet(ShareObjectViewSet):
-    serializer_class = serializers.TagSerializer
-    queryset = serializer_class.Meta.model.objects.all().select_related('extra')
-
-
-class SubjectViewSet(viewsets.ReadOnlyModelViewSet):
-    serializer_class = serializers.SubjectSerializer
-    queryset = serializer_class.Meta.model.objects.all()
-
-
-class ExtraDataViewSet(ShareObjectViewSet):
-    serializer_class = serializers.ExtraDataSerializer
-    queryset = serializer_class.Meta.model.objects.all()
-    filter_class = None
-
-
-# Agents
-
-class PersonViewSet(ShareObjectViewSet):
-    serializer_class = serializers.PersonSerializer
-    queryset = serializer_class.Meta.model.objects.all().select_related('extra')
-
-
-class OrganizationViewSet(ShareObjectViewSet):
-    serializer_class = serializers.OrganizationSerializer
-    queryset = serializer_class.Meta.model.objects.all().select_related('extra')
-
-
-class InstitutionViewSet(ShareObjectViewSet):
-    serializer_class = serializers.InstitutionSerializer
-    queryset = serializer_class.Meta.model.objects.all().select_related('extra')
-
-
-# Identifiers
-
-class AgentIdentifierViewSet(ShareObjectViewSet):
-    serializer_class = serializers.AgentIdentifierSerializer
-    queryset = serializer_class.Meta.model.objects.all().select_related('extra')
-
-
-class WorkIdentifierViewSet(ShareObjectViewSet):
-    serializer_class = serializers.WorkIdentifierSerializer
-    queryset = serializer_class.Meta.model.objects.all().select_related('extra')
-
-
-# Relations
-
-class AgentRelationViewSet(ShareObjectViewSet):
-    serializer_class = serializers.AgentRelationSerializer
-    queryset = serializer_class.Meta.model.objects.all().select_related('extra')
-
+        # Perform the lookup filtering.
+        lookup_url_kwarg = self.lookup_url_kwarg or self.lookup_field
 
-class WorkRelationViewSet(ShareObjectViewSet):
-    serializer_class = serializers.WorkRelationSerializer
-    queryset = serializer_class.Meta.model.objects.all().select_related('extra')
+        assert lookup_url_kwarg in self.kwargs, (
+            'Expected view %s to be called with a URL keyword argument '
+            'named "%s". Fix your URL conf, or set the `.lookup_field` '
+            'attribute on the view correctly.' %
+            (self.__class__.__name__, lookup_url_kwarg)
+        )
 
+        decoded_pk = IDObfuscator.decode(self.kwargs[lookup_url_kwarg])[1]
```

Comment:
Would it be possible to check that the model requested here is appropriate for this endpoint and then send back a 400 if it is not?

`decode` can also raise exceptions.  Could you make a custom exception for `InvalidID`, or something along those lines, and handle it here. Either a 400 or 404 would be fine, just depends on how nice you're feeling.